### DISCUSSION TITLE
:refer suggestions: getting it to work in Calva

### DISCUSSION THREAD

**Discussion by simon-brooke** [clj-kondo/clj-kondo#1307] [Ideas]
I spoke to [borkdude on Twotter](https://twitter.com/simon_brooke/status/1408746056712605702) this morning about getting Kondo to recommend a `:refer` vector when reporting 'unresolved symbol' from a namespace that had been required with `:refer :all`; and he replied that this should already work. However, this is currently not working for me in VS Code, and I'm wondering how I can make it do so.

As you can see in this picture, editing an unaltered copy of [this file](https://github.com/ertugrulcetin/jme-clj/blob/master/test/examples/beginner_tutorials/hello_asset.clj) 

![calva-refer](https://user-images.githubusercontent.com/5500763/123512323-bbce1280-d67e-11eb-883f-3d39da872327.png)

1. The installed version of Calva is 2.0.201, which is current latest;
2. The installed version of clj-kondo is 2021.6.18, which is current latest;
3. The version of Clojure NREPL support is 0.13.1, which is current latest;

When I run `clj-kondo --lint test/examples/beginner_tutorials/hello_asset.clj` from the command line I get:

```
test/examples/beginner_tutorials/hello_asset.clj:4:34: warning: use alias or :refer [add-light-to-root add-to-root attach-child bitmap-text box defsimpleapp detach-all-child geo get* gui-node light load-font load-model load-texture material re-init root-node rotate run scale set* start unbind-app vec3]
test/examples/beginner_tutorials/hello_asset.clj:20:9: warning: unused binding root-node
linting took 9ms, errors: 0, warnings: 2
```

which is exactly what I want. However, in Calva, I'm simply getting `unresolved symbol` warnings where the only suggested quick fix is `create private function 'attach-child'`, and, on the `(:require [jme-clj.core :refer :all])` expression, `use :alias or :refer ... (no quick fixes available)`.

Now, granted that this is clearly a problem not with `clj-kondo` but with how `calva` is handling its output, where would I start hacking to get Calva to present `:refer [add-light-to-root add-to-root attach-child bitmap-text box defsimpleapp detach-all-child geo get* gui-node light load-font load-model load-texture material re-init root-node rotate run scale set* start unbind-app vec3]` as a quick fix, either on the `(:require [jme-clj.core :refer :all])` expression or on one of the usages of symbols from that namespace?

Many thanks.

**Comment by borkdude** [comment]
This problem comes up quite often, but I don't know the downstream tooling problem yet. The suggestion that @ericdallo usually makes:

```
rm -rf .lsp/sqlite.db
```
and then restart VSCode. If this doesn't help, additionally, try removing `.clj-kondo/.cache` before doing the above. 

/cc @PEZ @ericdallo

**Reply by ericdallo** [reply]
I actually managed to reproduce the issue via cli on a project that has no `.clj-kondo` cache dir, what makes sense:
![image](https://user-images.githubusercontent.com/7820865/123517930-19139500-d67a-11eb-8b2f-0225f7c83e18.png)

After creating a `.clj-kondo` dir and running 
```bash
clj -Sdeps '{:deps {clj-kondo/clj-kondo {:mvn/version "2021.06.18"}}}' -m clj-kondo.main  --lint $(clj -Spath)
``` 
and then 

```bash
clj -Sdeps '{:deps {clj-kondo/clj-kondo {:mvn/version "2021.06.18"}}}' -m clj-kondo.main  --lint src/deps_bin/impl/bin.clj
```
I receive the expected lint:
```
src/deps_bin/impl/bin.clj:3:30: warning: use alias or :refer [repair-zip-with-preamble-bytes]                                                                                                 
linting took 176ms, errors: 0, warnings: 1 
```

So the issue is: You don't have a .clj-kondo dir properly on your project root so clj-kondo is not creating the cache or you have an outdated cache.

@borkdude I think I realized why it needs to remove `.lsp/sqlite.db`:
- clj-kondo will only fill the cache when scanning both namespaces which makes sense
- clojure-lsp will only trigger that when scanning the whole project, which happens one time during startup if doesn't contains a `.lsp/sqlite.db`

**Reply by ericdallo** [reply]
So removing `.lsp/sqlite.db` and creating `.clj-kondo`, it works on clojure-lsp:
![image](https://user-images.githubusercontent.com/7820865/123518284-94297b00-d67b-11eb-8d13-4b99c9da5fbb.png)

**Reply by borkdude** [reply]
@simon-brooke Can you confirm that this works for you?

**Reply by simon-brooke** [reply]
`.clj-kondo` already existed at the start of my investigation. I don't know whether or not it then had a `.cache` subdirectory. Hidden directories within hidden directories may be an unnecessary level of obfuscation!

So to start:

```
simon@mason:~/workspace/jme-clj$ ls -al .clj-kondo/
total 16
drwxrwxr-x  3 simon simon 4096 Jun  4 12:10 .
drwxrwxr-x 12 simon simon 4096 Jun 24 15:00 ..
drwxrwxr-x  6 simon simon 4096 Jun 26 12:31 .cache
-rw-rw-r--  1 simon simon  516 Jun  4 12:08 config.edn
ls -al .lsp
ls: cannot access '.lsp': No such file or directory
```

Interesting! I have not removed a `.lsp` subdirectory. Is there anywhere else `sqlite.db` could be hiding?

```
simon@mason:~/workspace/jme-clj$ rm -rf .clj-kondo/
```

(restart VSCode)
(Mouse over `(:require [jme-clj.core :refer :all])`)
![Screenshot from 2021-06-27 10-53-47](https://user-images.githubusercontent.com/5500763/123540342-681f0000-d736-11eb-9b1a-1f8d0498199c.png)

Not fixed.

However, the interesting question that that raises is: where is `sqlite.db` hiding and does it in fact exist at all?

**Reply by borkdude** [reply]
The fact that there is no `.lsp/sqlite.db` is a sign that clojure-lsp hasn't been analyzing your project, so this is a useful clue. 
Perhaps @ericdallo could comment on this.

**Reply by borkdude** [reply]
Also I think it would be useful to make a minimal repro project out of this so we can clone this and run it in Calva locally.

**Reply by simon-brooke** [reply]
Oh - I should add, the `.clj-kondo` folder which I deleted has also not been recreated.

```
simon@mason:~/workspace/jme-clj$ ls -l .clj-kondo
ls: cannot access '.clj-kondo': No such file or directory
simon@mason:~/workspace/jme-clj$ clj-kondo --lint "$(lein classpath)"
[extremely copious output]
linting took 9681ms, errors: 576, warnings: 1597
simon@mason:~/workspace/jme-clj$ ls -la .clj-kondo
ls: cannot access '.clj-kondo': No such file or directory
```

Still doesn't exist. Confused. Try again, doing what I did yesterday:

```
simon@mason:~/workspace/jme-clj$ clj-kondo --lint "$(lein classpath)"  --dependencies --parallel --copy-configs
[extremely copious output]
linting took 9458ms, errors: 576, warnings: 1597
simon@mason:~/workspace/jme-clj$ ls -la .clj-kondo
ls: cannot access '.clj-kondo': No such file or directory
```

OK, still doesn't exist. So, create it:

```
simon@mason:~/workspace/jme-clj$ mkdir .clj-kondo
simon@mason:~/workspace/jme-clj$ clj-kondo --lint "$(lein classpath)"  --dependencies --parallel --copy-configs
[extremely copious output]
linting took 9657ms, errors: 576, warnings: 1597
simon@mason:~/workspace/jme-clj$ ls -la .clj-kondo
total 12
drwxrwxr-x  3 simon simon 4096 Jun 27 11:10 .
drwxrwxr-x 12 simon simon 4096 Jun 27 11:10 ..
drwxrwxr-x  3 simon simon 4096 Jun 27 11:10 .cache
```

So: you have to create the `.clj-kondo` directory yourself, but once you have, `clj-kondo` will use it to hold a cache (I think @borkdude told me this yesterday).

Now again restart VS Code, and... no change. No appropriate quickfixes offered, either on `(:require [jme-clj.core :refer :all])` or on one of the symbols from that namespace (the unhelpful quickfix `Create provate function 'foo'` is still offered).

No progress, any further suggestions?

**Reply by simon-brooke** [reply]
Minimal repo is an excellent suggestion, will do this this evening.

**Reply by borkdude** [reply]
@simon-brooke 

>  I should add, the .clj-kondo folder which I deleted has also not been recreated.

You should make the `.clj-kondo` director yourself. This is the place where you store your config and this also signals to clj-kondo what is the root of your project. Please add it back and retry the above.

**Reply by borkdude** [reply]
Sorry, I didn't read the complete message before I responded. OK. And in Calva, no `.lsp` directory after you _restart_ VSCode?

**Reply by simon-brooke** [reply]
Exactly, yes.

**Reply by ericdallo** [reply]
No .LSP/sqlite.db is a bad sign, it means clojure-lsp is not analyzing all your project and issues like that could happen.
Please, check the clojjre-lsp log, you can the log path with Calva diagnostics command via the :log-path key

**Reply by ericdallo** [reply]
I guess there is even a log window on Calva for that log, not sure

**Reply by ericdallo** [reply]
I always pass the :log-path /tmp/clojjre-lsp.out so the log is always the same place to me

**Comment by ericdallo** [comment]
There is no code action for replacing a :all with it's used refers, at clojure-lsp we only have the clj-kondo warning suggesting that, so AFAICS I don't think it's a issue, but a feature suggestion

**Reply by ericdallo** [reply]
Actually, I never noticed that clj-kondo warning telling explictely what refers to use, but that probably would help setting the refers

**Reply by borkdude** [reply]
His issue is that the vars appear unresolved in Calva, but not with clj-kondo on the command line.

**Reply by simon-brooke** [reply]
In debugging this issue, I actually did see it for the first time - after having used Calva as my main editor for several months. This implies that, although the correct behaviour can happen, it's pretty rare.

**Reply by borkdude** [reply]
@simon-brooke ok, a repro (a git repo we could actually clone) would help @ericdallo, @PEZ and me diagnose the issue locally with Calva.

**Reply by simon-brooke** [reply]
OK, I can now demonstrate it. Check out this branch: https://github.com/simon-brooke/kondo-hint-investigation/tree/demo (i.e., the `demo` branch). If I've committed all the right bits to git, when you open `kondo-hint-investigation.core` in Visual Studio Code using Calva, you'll see the following namespace declaration:

```clojure
(ns kondo-hint-investigation.core
  (:require [clojure.data.json :refer :all]
            [jme-clj.core :refer :all]))
```

If you mouse over the keyword `:all` in `[clojure.data.json :refer :all]`, you'll get the hint `use alias or :refer [write-str]`, which is the right behaviour. If you mouse over the keyword `get` in `[jme-clj.core :refer :all]` you get just `use alias or refer`, which is the wrong behaviour.

I suspect that the difference is that `clojure-lsp` is exploring for dependencies first time it encounters a new namespace, but does not explore for dependencies subsequently. Note, that's a hypothesis, I don't know it to be true.

**Reply by ericdallo** [reply]
It works good to me with lsp-mode (Emacs): 
![image](https://user-images.githubusercontent.com/7820865/123639536-b09df280-d7f6-11eb-91f1-1d5eae90ffdf.png)
This is the clojure-lsp log I get:
![image](https://user-images.githubusercontent.com/7820865/123639584-babff100-d7f6-11eb-893e-d4d1d2625b7b.png)

**Reply by borkdude** [reply]
I had to remove the `.clj-kondo/.cache` and `.lsp/sqlite.db` from the git repo. These aren't things you're supposed to check into source control. But after doing so and re-opening VSCode, it shows up correctly for me.

<img width="1288" alt="Screenshot 2021-06-28 at 15 01 42" src="https://user-images.githubusercontent.com/284934/123640865-e2787e80-d821-11eb-83c5-d93fa831564c.png">

**Reply by simon-brooke** [reply]
Exactly, that's *why* I checked `.clj-kondo/.cache` and `.lsp/sqlite.db` into the git repo - so that you could see exactly the behaviour that I had. If you delete them, then it reinitialises and gets it right. But if you're developing a project and decide you need a new library, then (I think) `clojure.lsp` does not run a full scan on dependencies, and consequently you don't get the appropriate hint on mouseover.

It's a small detail, I know, but that's typically how people do develop things. They don't start a new project and immediately know all the libraries it's going to need; rather, they add dependencies as they discover that they need them. So probably, the right time to run a full dependencies scan is each time the `project.clj` is saved.

**Reply by borkdude** [reply]
> But if you're developing a project and decide you need a new library

This is up to clojure-lsp and/or Calva. @ericdallo any thoughts? I think clojure-lsp should maybe hash deps.edn and invalidate some cache when it changes?

**Reply by ericdallo** [reply]
I could not repro the issue even with your committed db, not sure why this is happening for you.

> This is up to clojure-lsp and/or Calva. @ericdallo any thoughts? I think clojure-lsp should maybe hash deps.edn and invalidate some cache when it changes?

clojure-lsp already `(disgest/md5 project-file)`, so any changes to project.clj would trigger a new analysis, so @simon-brooke your suggestion [already happens](https://github.com/clojure-lsp/clojure-lsp/blob/master/src/clojure_lsp/crawler.clj#L47).

**Reply by ericdallo** [reply]
I tested multiple ways that could cause any harm, I checked out main after demo, tested not removing cache files and removing it, but I could not repro your issue.

**Reply by simon-brooke** [reply]
Thank you very much for your effort, @ericdallo. I shall try it on other machines, to see it I have a setup issue.
